### PR TITLE
Add footer

### LINKED ISSUE #515: Feature: Footer
Footer to include: 

Social Links
- [ ] GitHub
- [ ] Twitter
- [ ] newsletter call-out?
- [ ] Email? currently we push to "Get in touch" on contribute so they'll try GH or whatever's appropriate first

Related projects: 
- [ ] IPFS
- [ ] Filecoin
- [ ] PL
- [ ] IPLD ? 
- [ ] Multiformats ? 

Feedback mechanisms (see also #426) 
Currently we're using 2 separate footers within tutorials - One to collect feedback on specific lessons and another to collect feedback on specific tutorials. We also have the 3-question tutorial satisfaction survey at the end of each tutorial, followed by a link to the user survey. It's important that if we we remove the current footer versions of lesson/tutorial feedback that we still have some kind of "Feeling stuck?" etc to prompt for GH submission re both specific lessons and overall tutorials. 

Do we want a bit of about-style text here that would add to our character counts for real content on each page?

### PR DESCRIPTION
**Pull request by terichadbourne**
WIP - will close #515 by adding footer (both for IA and SEO purposes) 

image

Note that the callout that looked like a navy footer requesting feedback on specific lessons has been changed to aqua so it can be used in combo with this new universal footer where needed...

image

This needs more styling work and more discussion about what content to include, including... 

**Styling improvements needed**
- [x]  Improve logo placement
- [x]  Center copyright notice?
- [x]  Improve vertical spacing in lists
- [x]  Change styling of list headers (for SEO purposes, should we be using <h*> tags at the top of each list or just styled text?)
- [x] Confirm width of header, footer, and body make sense together

**Content decisions to discuss**
- [ ] Any reason to duplicate things that are easily accessible from top nav? (see more content suggestions from @zebateira in #515) 
- [ ] Way to call out newsletter signup more prominently? 
- [x] Do we need any "about" type text either to highlight what ProtoSchool is or to bulk up amount of content on each page for SEO? (The Filecoin website includes its basic self-description in the footer, but we already have that in our non-sticky header.) 
- [ ] It's weird that courses and related projects list are so similar to each other, but courses aren't flagged anywhere else internally and related projects may be our biggest SEO improvement. 

**Bugs**
- [x]  Link to contact section of contribute page isn't currently working - need to confirm we can do this in Vue now that we use history instead of hash routing

**Other**
- [x] Set up as text strings from translation JSON
- [x] Any tests needed?

image

@zebateira Please feel free to add any other concerns/questions to the list above. I know you have some styling suggestions in mind... why don't you jot those down here and I can tackle or we can work together during a pairing session so I get the experience making CSS bend to my will.

### CODE REVIEW

**Comment by zebateira** [comment]
Some suggestions:


- [ ] Remove top level div from `Footer.vue` component
- [ ] Add extra margin to the top of the footer (2rem should be ok)
- [ ] Change padding to `pv4` on main footer container
- [ ] Add missing `ph3` to the content containers (`footer > div`) so it's aligned with the rest of the app
- [ ] Drop `justify-between` to space out the columns and use `w-25` instead
- [ ] Add `pv1` on each `li`
- [ ] Add simple glow effect for each `li` by adding `o-80 glow` to `a`
- [ ] Add `mt0` to each list header
- [ ] Add `ProtoSchool` header title and the header slogan next to the logo (text should break downwards aligning with the header title; ideal logo width likely around `w2`)
- [ ] Maybe align copyright notice to the right?
- [ ] Mobile
      - [ ] break the navigation downwards and see if keeping the grid at 3 columns works, if not reduce to 2 columns.

**Comment by terichadbourne** [comment]
Made some additional styling tweaks based on feedback from design. 

@zebateira Is there anything in this PR that feels worthy of me adding tests for? I assume we could easily add tests for internal links rendering the right pages, but perhaps that's overkill? Looks like Cypress intentionally won't test new windows that you've opened (for our external pages), but can test whether a window.open event happened. https://stackoverflow.com/questions/47749956/access-a-new-window-cypress-io

**Comment by zebateira** [comment]
> Made some additional styling tweaks based on feedback from design.
> 
> @zebateira Is there anything in this PR that feels worthy of me adding tests for? I assume we could easily add tests for internal links rendering the right pages, but perhaps that's overkill? Looks like Cypress intentionally won't test new windows that you've opened (for our external pages), but can test whether a window.open event happened. https://stackoverflow.com/questions/47749956/access-a-new-window-cypress-io

It would nice to have the tests for the external pages so that we could avoid having out of date links in the future, but at the same time it would be a problem because it would skew analytics for the those websites.
I'd say skip testing - maybe just add one or two tests for checking the footer is being rendered in the homepage and some other page.

File: src/components/Lesson.vue [hunk 1]
```diff
@@ -142,7 +145,8 @@ import {
 } from '../utils/tutorials'
 import countly from '../utils/countly'
 import marked from '../utils/marked'
-import Header from './Header.vue'
+import Header from './layout/Header.vue'
+import Footer from './layout/Footer.vue'
```
Comment (zebateira) [inline src/components/Lesson.vue:149]: 💯

File: src/components/layout/Footer.vue [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      return translations.footer
+    },
+    processedColumns: function () {
+      return translations.footer.columns.map(column => ({
+        ...column,
+        links: column.links.map(link => {
+          let text = link.text // text or undefined
+          let url = link.url // url or underfined
+
+          if (column.type === 'courses' || column.type === 'projects') {
```
Comment (zebateira) [inline src/components/layout/Footer.vue:52]:
```suggestion
          // Fill in course and project info link data automatically
          if (column.type === 'courses' || column.type === 'projects') {
```

File: src/components/layout/Footer.vue [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            let project = projects.find(project => project.id === link)
+
+            text = project.name
+            url = (column.type === 'courses') ? `/course/${link}` : project.url
+          }
+
+          return { text, url, external: !(url.startsWith('/')) }
```
Comment (zebateira) [inline src/components/layout/Footer.vue:59]:
```suggestion
          return { text, url, external: !url.startsWith('/') }
```

**Review by zebateira — CHANGES_REQUESTED** [review]
👌 👌 👌

File: src/components/layout/Footer.vue (hunk 2 above)
Comment (terichadbourne) [reply src/components/layout/Footer.vue:60]: @zebateira You're just suggesting adding the comment here, not suggesting that I do something we haven't done yet?

Comment (zebateira) [reply src/components/layout/Footer.vue:60]: Sorry yes. Just to make it clear of the purpose of processing these columns. Makes sense?

**Comment by terichadbourne** [comment]
@zebateira I've removed the failing test for external links as we discussed, since we can't catch the window.open event in the way we'd hoped. Mind giving this a final review on Monday? (Please jog my memory if I've forgotten about any other last changes we discussed.)

**Review by zebateira — APPROVED** [review]

**Review by zebateira — APPROVED** [review]
LGTM 💯

**Comment by zebateira** [comment]
@terichadbourne should we create another issue about these things here that we didn't decide on?

#### Content decisions to discuss
- [ ] Any reason to duplicate things that are easily accessible from top nav? (see more content suggestions from @zebateira in #515)
- [ ] Way to call out newsletter signup more prominently?
- [ ] It's weird that courses and related projects list are so similar to each other, but courses aren't flagged anywhere else internally and related projects may be our biggest SEO improvement.

**Pull request merged**